cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java

```diff
@@ -0,0 +1,359 @@
+      Executors.newCachedThreadPool(Threads.createDaemonThreadFactory("notification-subscriber-task")));
+    taskRunner = new ScheduleTaskRunner(store, lifecycleService, propertiesResolver,
+                                        taskExecutorService, namespaceQueryAdmin, cConf);
+  }
+
+  @Override
+  protected void run() {
+    LOG.info("Start running NotificationSubscriberService");
+    if (!isRunning()) {
+      return;
+    }
+
+    subscriberThreads.add(new TimeEventNotificationSubscriberThread(null));
+    subscriberThreads.add(new DataEventNotificationSubscriberThread(null));
+
+    for (NotificationSubscriberThread thread : subscriberThreads) {
+      thread.start();
+    }
+
+    for (NotificationSubscriberThread thread : subscriberThreads) {
+      try {
+        thread.join();
+      } catch (InterruptedException e) {
+        LOG.info("Thread {} is being terminated while waiting for it to finish.", thread.getName());
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  @Override
+  protected void triggerShutdown() {
+    LOG.info("Stopping NotificationSubscriberService.");
+    for (NotificationSubscriberThread thread : subscriberThreads) {
+      thread.interrupt();
+    }
+
+    LOG.info("NotificationSubscriberService stopped.");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (taskExecutorService != null) {
+      taskExecutorService.shutdownNow();
+    }
+  }
+
+  private abstract class NotificationSubscriberThread extends Thread {
+    private final String topic;
+    private final RetryStrategy scheduleStrategy;
+    protected final Deque<Job> readyJobs;
+    private int emptyFetchCount;
+    private int failureCount;
+    private String messageId;
+
+
+    NotificationSubscriberThread(String topic, @Nullable String messageId) {
+      super(String.format("NotificationSubscriberThread-%s", topic));
+      this.topic = topic;
+      this.messageId = messageId;
+      // TODO: [CDAP-11370] Need to be configured in cdap-default.xml. Retry with delay ranging from 0.1s to 30s
+      scheduleStrategy =
+        co.cask.cdap.common.service.RetryStrategies.exponentialDelay(100, 30000, TimeUnit.MILLISECONDS);
+      this.readyJobs = new ArrayDeque<>();
+    }
+
+    @Override
+    public void run() {
+      while (isRunning()) {
+        try {
+          long sleepTime = processNotifications();
+          // Don't sleep if sleepTime returned is 0
+          if (sleepTime > 0) {
+            TimeUnit.MILLISECONDS.sleep(sleepTime);
+          }
+        } catch (InterruptedException e) {
+          // sleep is interrupted, just exit without doing anything
+        }
+      }
+    }
+
+    /**
+     * Fetch new notifications and update job queue
+     *
+     * @return sleep time in milliseconds before next fetch
+     */
+    private long processNotifications() {
+      try {
+        final MessageFetcher fetcher = messagingContext.getMessageFetcher();
+        emptyFetchCount++;
```

Comment:
It seems strange that you increment the empty fetch count before checking if it is empty. The logic might be correct, but not easy to read.